api/bucketsd/service.go

```diff
@@ -961,7 +1007,64 @@ func (s *Service) createLinks(
 	}, nil
 }
 
-func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (res *pb.SetPathResponse, err error) {
+func (s *Service) copyNode(ctx context.Context, buck *tdb.Bucket, rootNode ipld.Node, fromPath string, toPath string) (path.Resolved, error) {
+	fileKey, err := buck.GetFileEncryptionKeyForPath(toPath)
+	if err != nil {
+		return nil, err
+	}
+
+	currentFileKeys, err := buck.GetFileEncryptionKeysForPrefix(fromPath)
+	if err != nil {
+		return nil, err
+	}
+	newFileKeys, err := buck.GetFileEncryptionKeysForPrefix(toPath)
+	if err != nil {
+		return nil, err
+	}
+	nmap, err := s.encryptDag(
+		ctx,
+		rootNode,
+		fromPath,
+		buck.GetLinkEncryptionKey(),
+		currentFileKeys,
+		newFileKeys,
+		fileKey,
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+	nodes := make([]ipld.Node, len(nmap))
+	i := 0
+	for _, tn := range nmap {
+		nodes[i] = tn.node
+		i++
+	}
+	pn := nmap[rootNode.Cid()].node
+	ctx, dirPath, err := s.insertNodeAtPath(ctx, pn, path.Join(path.New(buck.Path), toPath), buck.GetLinkEncryptionKey())
+	if err != nil {
+		return nil, fmt.Errorf("updating pinned root: %v", err)
+	}
+
+	_, err = s.addAndPinNodes(ctx, nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	// If updating root, add seedfile back to node.
+	if toPath == "" {
+		sn, err := makeSeed(fileKey)
+		ctx, dirPath, err = s.insertNodeAtPath(ctx, sn, path.Join(dirPath, buckets.SeedName), buck.GetLinkEncryptionKey())
```

Comment:
Do we need to pin it as well (`addAndPinNodes`)?

Reply:
will bump the one at 1049 below and append nodes if this section happens.